### ISSUE TITLE
[DRAFT] Prefetch the advertising info 

### ISSUE THREAD

**Issue by YuriyVelichkoPI** [prebid/prebid-mobile-android#696]
## Background

In the current implementation, the SDK requests the [advertizing info](https://developers.google.com/android/reference/com/google/android/gms/ads/identifier/AdvertisingIdClient#getAdvertisingIdInfo(android.content.Context)) on each bid request. To avoid blocking it is performed async in the background thread. However, the  Prebid SDK in any case is waiting until the OS returns the current advertising info. See the class `AdIdManager` of the SDK for the details. 

This implementation has two major disadvantages: 
- Produces runtime errors due to the complexity of the async behavior and Android behavior. 
- Slows the bidding process that can be critical to publishers with low average app session time. 

To avoid runtime issues and get rid of delays in creating the bid request, the SDK should prefetch the advertising info so it's always available. The prefetching should take into consideration the [Android policy ](https://developer.android.com/training/articles/ad-id)to request the updated info each time it's needed and not cache it. 

## Proposal

Even though Android docs advise not to cache the ID it's clear that this property won't be changed too often, especially during the single app session. So, if the SDK will fetch it on the initialization and refresh on each following bid request, it shouldn't violate privacy settings and other policies. 

The general approach is displayed in the following diagram:

![Prebid - Github Proposals](https://github.com/prebid/prebid-mobile-android/assets/21169695/c36b9d7d-ec09-40e8-af8e-2ab52e651b66)

Step 1: The SDK on the initialization step runs the background task to fetch the advertising info and save it in memory. No cache on disc. 
Step 2: Once the SDK is initialized, publishers can run bid requests
Step 3: Publisher calls the `fetchDemand` method, or any other, that leads to the creation and sending of the Bid Request.
Step 4: Prebid SDK uses the prefetched advertising info to build the Bid Request
Step 5: Prebid SDK updates the advertising info, starting the respective background task. The SDK doesn't wait for the result. 
Step 6: SDK makes the bid request with prefetched advertising info 

Once the updating task is finished, the actual advertising info will be updated and used on the next bid request. 

Note: the diagram and description don't contain, but assume that fetching the advertising info is performed according to the user consent and permission as it is happening now. 

## Objectives

Once the proposal is reviewed, finalized, and approved, the implementation will assume the following steps:
- Add the advertising ID fetching task to the SDK initialization 
- Update the flow of the building the Bid Request:
  - Use the prefetched advertising info
  - Run the async task to update the advertising info
- Make sure that updating of the advertising info is not performed too often. If the publisher runs bid requests for several ad units at the same time, there is no need to run an updating task for each request. We can introduce policies like "run no more than one advertising update task per minute". Hence the method `AdIdManager.updateAdvertizingId()` should check the last update time before starting a new background task. 
- Cover new behavior with unit tests 
- Prepare run and study benchmarks for the `initilizeSDK` method to understand the impact of the new task.

**Comment by YuriyVelichkoPI** [comment]
History note. The motivation for the proposed changes was raised in the discussion of [this](https://github.com/prebid/prebid-mobile-android/pull/692) PR. This is not the first time we have experienced problems with the sequential retrieval of the advertising info while preparing the bid request.
